File: coding/src/com/algos/li13_graph/revised/Dfs_L1020_NumberOfEnclaves.java
package  com.algos.li13_graph.revised;

import java.util.LinkedList;
import java.util.Queue;

public class Dfs_L1020_NumberOfEnclaves {
    public static void main(String[] args){
        Dfs_L1020_NumberOfEnclaves obj = new Dfs_L1020_NumberOfEnclaves();

        int grid[][] = {
        {0, 0, 0, 0},
        {1, 0, 1, 0},
        {0, 1, 1, 0},
        {0, 0, 0, 0}};

        int ans = obj.numEnclaves(grid);
        System.out.println(ans);
    }  
    
    // Problem: https://leetcode.com/problems/number-of-enclaves/
    public int numEnclaves(int[][] grid) {
        int m = grid.length, n = grid[0].length; 

        int vis[][] = new int[m][n]; 
        for(int i = 0;i<m;i++) {
            for(int j = 0 ; j<n;j++) {
                if(i == 0 || j == 0 || i == m-1 || j == n-1) 
                    if(vis[i][j] == 0 && grid[i][j] == 1) 
                        dfs(i, j, vis, grid); 
            }
        }
        
        int count = 0;
        for(int i=0; i<m; i++) {
            for(int j=0;j<n;j++) {
                if(grid[i][j] == 1 && vis[i][j] == 0) 
                    count++; 
            }
        }
        return count;         
    }

    private void dfs(int row, int col,int vis[][], int grid[][]) {
        if (!isSafe(vis, grid, row, col)) 
            return;     

        vis[row][col] = 1; 
        dfs(row+1, col, vis, grid);
        dfs(row-1, col, vis, grid);
        dfs(row, col+1, vis, grid);
        dfs(row, col-1, vis, grid);
    }

    private boolean isSafe(int[][] vis, int[][] grid, int nrow, int ncol) {
        int m = grid.length, n = grid[0].length;
        return nrow >= 0 && nrow < m && ncol >= 0 && ncol < n
                && vis[nrow][ncol] == 0 && grid[nrow][ncol] == 1;
    }
}

File: dsa/patterns/13_graphs/python/dfs_l1020_number_of_enclaves.py
"""
Problem: https://leetcode.com/problems/number-of-enclaves/

Idea:
- Any land cell (1) connected to the boundary cannot be an enclave
- Run DFS from all boundary land cells and mark them as visited
- Finally count all land cells which are NOT visited
- Those are enclaves

Time: O(m * n)
Space: O(m * n)
"""


def num_enclaves(grid):
    m, n = len(grid), len(grid[0])
    visited = [[0] * n for _ in range(m)]

    # Run DFS from all boundary cells
    for i in range(m):
        for j in range(n):
            if i == 0 or j == 0 or i == m - 1 or j == n - 1:
                if grid[i][j] == 1 and visited[i][j] == 0:
                    dfs(grid, i, j, visited)

    # Count remaining unvisited land cells
    count = 0
    for i in range(m):
        for j in range(n):
            if grid[i][j] == 1 and visited[i][j] == 0:
                count += 1

    return count


def dfs(grid, row, col, visited):
    if not is_safe(grid, row, col, visited):
        return

    visited[row][col] = 1

    dfs(grid, row + 1, col, visited)
    dfs(grid, row - 1, col, visited)
    dfs(grid, row, col + 1, visited)
    dfs(grid, row, col - 1, visited)


def is_safe(grid, row, col, visited):
    m, n = len(grid), len(grid[0])
    return (
        0 <= row < m and
        0 <= col < n and
        grid[row][col] == 1 and
        visited[row][col] == 0
    )


# -------------------- TEST --------------------
if __name__ == "__main__":
    grid = [
        [0, 0, 0, 0],
        [1, 0, 1, 0],
        [0, 1, 1, 0],
        [0, 0, 0, 0]
    ]

    print(num_enclaves(grid))
